[WIP] build: add experimental watch mode

bajtos opened the pull request:
**Let's make work in Lerna+Typescript monorepo fun again!**

In this pull request, I am introducing an experimental `tsc --watch` setup that works in a monorepo 🎉 

The problem: `tsc` does not respect the layout of our monorepo and writes compiled output to `{outDir}/{packageName}/{localPath}` instead of the desired `packages/{packageName}/dist/{localPath}`.

The solution: let `tsc` write the files to a hidden `.build` directory. Write a custom file watcher that picks any files changed by `tsc` in `.build` and copies them to their correct place in `packages` hierarchy.

In addition to this wrapper, this pull request also:

- Adds a new package script: `npm run build:watch`
- Add two new VSCode tasks and configure them as the default Build Task and Test Task respectively:
  - Watch and compile source files
  - Run all tests
- Configures prettier to ignore ALL json files (so that `.vscode/tasks.json` stays formatted as we like it)
- Upgrades VSCode tasks.json to version 2.0.0

Time to run `npm run build`: 21s
Time to finish the first compilation via `npm run build:current`: 4s
Time to recompile after changing a file: <1s
Time to run `npm run mocha` (for comparison): 7s

_I am fully aware of the fact that this work was not planned for February milestone. However, the current setup where the build takes so long leaves me unable to contribute any non-trivial code changes. I think the time needed to make this change happen (including reviews) will pay back very quickly._

<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->


## Checklist

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] Related API Documentation was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `packages/example-*` were updated

raymondfeng commented on package.json:
```diff
@@ -32,10 +36,11 @@
     "prettier:check": "npm run prettier:cli -- -l",
     "prettier:fix": "npm run prettier:cli -- --write",
-    "clean": "lerna run clean --loglevel=silent --parallel",
+    "clean": "rimraf .build && lerna run clean --loglevel=silent --parallel",
```
No need to use `rimraf` as an extra dependency. Let's stay with `node packages/build/bin/run-clean`.

virkt25 commented on .vscode/tasks.json:
```diff
@@ -1,41 +1,44 @@
     },
     {
-      "taskName": "Run all tests",
+      "label": "Build and run and lint",
```
Can we call this `Build, test and lint`.

virkt25 commented on .vscode/tasks.json:
```diff
@@ -1,41 +1,44 @@
-      "taskName": "Install all dependencies",
+      "label": "Install all dependencies",
+      "group": "build",
       "command": "./node_modules/.bin/lerna",
```
Maybe we should switch to `npm` commands so this is same as what a developer does today? Also helps ensure consistency w.r.t. those scripts changing, this won't have to be updated. A good example is that this is missing the `--parallel` optimization found in the npm script.

```
command: "npm"
args: ["run", "bootstrap"]
```

Comment applies to commands below using `./node_modules/.bin/lerna` as well.

virkt25 commented on .vscode/tasks.json:
```diff
@@ -1,41 +1,44 @@
     },
     {
-      "taskName": "Remove all dependencies",
+      "label": "Remove all dependencies",
```
Since there is a task below to `Clean all generated files`, do we need this task? I personally just use `npm run clean` ... which does the same as task below.

bajtos replied:
We have two different tasks:

 - clean files generated by our build (`dist`, `api-docs`, etc.) - this is `npm run clean`
 - remove `node_modules` in all packages - this is `npm run clean:lerna`

virkt25 commented on .vscode/tasks.json:
```diff
@@ -114,13 +117,45 @@
     {
-      "taskName": "Clean all generated files",
-      "isShellCommand": true,
+      "label": "Clean all generated files",
```
Can we call this `Clean project`? This is a suggestion not a blocker. (I prefer simple names).

virkt25 commented on .vscode/tasks.json:
```diff
@@ -114,13 +117,45 @@
+      },
+      "type": "shell",
+      "command": "npm",
+      "args": ["run", "mocha", "-s"],
```
Should linting be a part of this? Normally covered by `posttest` but we can't trigger `test` here since that builds and this is for watch right?

bajtos replied:
Yes, this is for watch, thus we don't want to trigger `pretest` hook.

I am also intentionally not running any linter from this tasks, because typically linting is provided on the fly by Prettier/TSLint extensions installed in VS Code.

Sure, some linting problem won't be caught by those plugins (e.g. tslint rules requiring type info). To me, it's more important to have a fast red-green-refactor cycle that may miss some issues than to have a slow red-green-refactor cycle that always catches everything.

Here the workflow I am imagining:

 1. Open VSCode
 2. Start the watching build in background
 3. Write a failing tests, quickly run tests to see the new test fail (no linters)
 4. Make the new test pass, quickly run tests to verify
 5. Refactor, run tests to ensure nothing broke
 6. Rinse and repeat steps 2-5
 7. Run `npm test` from the command line
 8. Commit the changes

Note that our current watching build is not emitting `.d.ts` files, therefore it will not report missing imports needed by type definition files only (see e.g. https://github.com/Microsoft/TypeScript/issues/21673) - another reason why we should not rely on this "quick" mode and run `npm test` from time to time.

virkt25 commented on bin/tsc-watch-monorepo.js:
```diff
@@ -0,0 +1,83 @@
+const build = require('../packages/build');
+const debug = require('debug')('build-watch');
+const chokidar = require('chokidar');
+const fs = require('fs');
```
Any reason to not use `fs-extra`? We can use `ensureDir` (also aliased to `mkdirp`) so we don't need the below dependency. Plus async/await support.

virkt25 commented on bin/tsc-watch-monorepo.js:
```diff
@@ -0,0 +1,83 @@
+const MONOREPO_DIR = path.resolve(__dirname, '..');
+const BUILD_DIR = path.resolve(MONOREPO_DIR, '.build');
+
+mkdirp.sync(BUILD_DIR);
```
`fs.mkdir()` would do the same thing ... this isn't a nested folder. But see comment above about using `fs-extra`

virkt25 commented on tsconfig.json:
```diff
@@ -1,5 +1,11 @@
+    "node_modules/**",
+    "packages/*/node_modules/**",
+    "**/*.d.ts",
+    "packages/*/test/sandbox/**",
```
I thought we were switching to `packages/*/.sandbox` for the recommended TestSandbox Path within the project?

bajtos replied:
AFAICT, this change has not happened yet.

raymondfeng commented:
@bajtos The following command functionally does the same job by watching each of the packages in parallel.

`lerna run --parallel build -- -- --watch`

Please note `-- -- --watch` allows us to pass `--watch` to `"build"` script for each package. It basically runs `lb-tsc es2017 --watch` for all packages. 

What do you think?

bajtos commented:
> The following command functionally does the same job by watching each of the packages in parallel.
> It basically runs `lb-tsc es2017 --watch` for all packages.

As far as I understand the watch mode, it does not watch dependencies in `node_modules` for changes. So for example, when we make a change in `@loopback/context`, the `tsc` process watching files in `packages/core` will not rebuild affected files. I did a quick check now and it looks like this is actually not true. When I renamed `export namespace inject` to `export namespace injectX`, several downstream builds were triggered for me. This is good!

Few problems remain though: because we are running multiple builds in parallel, there is no easy way how to tell when the build has finished. VS Code uses regular expression to detect lines like "File change detected" and "Compilation complete" - see [Background/watching tasks](https://code.visualstudio.com/docs/editor/tasks#_background-watching-tasks). This detection does not work with your approach OOTB. When I fixed the regular expression used by `$tsc-watch` problem matcher to allow `@loopback/{name}: ` prefix added by lerna, VSCode interpreted the first matching message as the end of the build process, even though there were other builds still running. 

Also note that `tsc --watch` cleans the screen before starting a new build, this does not work well when there are multiple `tsc` processes cleaning the same shared screen.

I suppose that instead of writing my tsc-watch-monorepo script, we could write a script that would aggregate output of multiple `tsc` processes running under the hood and clean it for consumption by VSCode (or any other tooling that consumes tsc watch output). I am not sure if such script would be a better solution, especially in regards to code complexity.

---

Hmm. As I was playing with these two different approaches, I found that my solution actually does not work. When I renamed `inject` namespace to `injectX`, no error was reported by the compiler. It looks like my special watch mode does not recognize dependencies between our packages and thus does not recompile consumers of `@loopback/context` module.

Back to the drawing board!

bajtos commented:
Thank you all for your feedback! Looks like this patch will require more time to get right. Let's start by landing changes that are not tied to tsc-watch - see #987. I'll keep this pull request open as a longer-term work and let you know when it's ready for review again.

bajtos commented:
I don't have bandwidth to work on this feature in the near future and since it's not part of our roadmap, I think it's best to close this pull request for now.
